File: progressive_deploy/A5-endpoint/js/src/ip-guard.js
function ipToInt(ip) {
  const parts = ip.split('.');
  return ((parseInt(parts[0], 10) << 24) |
    (parseInt(parts[1], 10) << 16) |
    (parseInt(parts[2], 10) << 8) |
    parseInt(parts[3], 10)) >>> 0;
}

function ipInCIDR(ip, cidr) {
  const [net, bitsStr] = cidr.split('/');
  const bits = parseInt(bitsStr, 10);
  const mask = ~((1 << (32 - bits)) - 1);
  return (ipToInt(ip) & mask) === (ipToInt(net) & mask);
}

let _blacklist = null;

function loadBlacklist(env) {
  const raw = env.IP_BLACKLIST || '';
  if (!raw) return [];
  return raw.split(',').map((s) => s.trim()).filter(Boolean);
}

export function isIPBlocked(ip, env) {
  if (!ip) return false;
  if (_blacklist === null) {
    _blacklist = loadBlacklist(env);
  }
  for (const entry of _blacklist) {
    if (entry.includes('/')) {
      if (ipInCIDR(ip, entry)) return true;
    } else if (ip === entry) {
      return true;
    }
  }
  return false;
}

File: src/skeleton/endpoint/A5-endpoint/python/core/ip_guard.py
import os
import ipaddress
import logging

logger = logging.getLogger(__name__)

_blacklist = None


def _load_blacklist():
    raw = os.getenv("IP_BLACKLIST", "")
    if not raw:
        return []

    networks = []
    for item in raw.split(","):
        item = item.strip()
        if not item:
            continue
        try:
            networks.append(ipaddress.ip_network(item, strict=False))
        except ValueError:
            logger.warning("Invalid IP blacklist entry: %s", item)

    return networks


def _get_blacklist():
    global _blacklist
    if _blacklist is None:
        _blacklist = _load_blacklist()
    return _blacklist


def is_ip_blocked(ip_str: str) -> bool:
    try:
        ip = ipaddress.ip_address(ip_str)
    except ValueError:
        logger.warning("Cannot parse client IP: %s", ip_str)
        return False

    for net in _get_blacklist():
        if ip in net:
            logger.warning("Blocked IP: %s (matched %s)", ip_str, net)
            return True

    return False
